File: src/java/DemoWeb/src/com/csoftz/demo/web/ProductController.java
/*----------------------------------------------------------------------------*/
/* Source File:   PRODUCTCONTROLLER.JAVA                                      */
/* Description:   Controller class for product page                           */
/* Author:        Carlos Adolfo Ortiz Quir—s (COQ)                            */
/* Date:          Jan.04/2011                                                 */
/* Last Modified: Jan.07/2011                                                 */
/* Version:       1.3                                                         */
/* Copyright (c), 2011 CSoftZ                                                 */
/*----------------------------------------------------------------------------*/
/*-----------------------------------------------------------------------------
 History
 Jan.04/2011 COQ  File created.
 -----------------------------------------------------------------------------*/

package com.csoftz.demo.web;

import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;

import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import com.csoftz.demo.common.consts.GlobalConstants;
import com.csoftz.demo.common.consts.RequestConstants;
import com.csoftz.demo.domain.ProductInfo;
import com.csoftz.demo.domain.search.ProductSearchInfo;
import com.csoftz.demo.intr.data.IProductDataService;

/**
 * Controller class for home page (HOME.JSP)
 * 
 * @since 1.5(JDK), Jan.04/2011
 * @author Carlos Adolfo Ortiz Quir—s (COQ)
 * @version 1.3, Jan.07/2011
 */
@Controller
@RequestMapping(value = "/product")
public class ProductController extends AbstractComponentBaseController
		implements InitializingBean {

	/**
	 * This is injected by IoC. Access to products.
	 */
	private IProductDataService productService = null;

	/**
	 * This is the welcome page or index page for Products.<br/>
	 * Responds to the URI GET: {webapp}/product
	 * 
	 * @param model
	 *            Values to pass through to view.
	 * @return View name to render
	 */
	@RequestMapping(value = "", method = RequestMethod.GET)
	public String index(Model model, HttpServletRequest request) {
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product request. Index Action");
		}
		List<ProductInfo> l = productService.getAll();
		model.addAttribute("productSearchInfo", new ProductSearchInfo());
		model.addAttribute("productList", l);
		model.addAttribute("productListEmpty", l.isEmpty());
		
		String msgBox = (String) request.getSession(true).getAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG);
		request.getSession(true).removeAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG);
		if (msgBox == null) {
			msgBox = "";
		}
		model.addAttribute(RequestConstants.REQUEST_HTTP_MSG_BOX, msgBox);
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Index Action.");
		}
		return "product/index";
	}

	/**
	 * This is the entry point to creating new record.<br/>
	 * Responds to the URI GET: {webapp}/product/create
	 * 
	 * @param model
	 *            Values to pass through to view.
	 * 
	 * @return View name to render
	 */
	@RequestMapping(value = "/create", method = RequestMethod.GET)
	public String create(Model model) {
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product/create request. Create Action.");
		}
		model.addAttribute(RequestConstants.REQUEST_HTTP_FORM_COMMAND,
				RequestConstants.REQUEST_HTTP_FORM_COMMAND_CREATE);
		model.addAttribute("productInfo", new ProductInfo());
		model.addAttribute("message", "Creating new record");
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Create Action.");
		}
		return "product/edit";
	}

	/**
	 * This is the entry to edit a record.<br/>
	 * Responds to the URI GET: {webapp}/product/edit/{id}. Notice: It is passed
	 * an id to work on.
	 * 
	 * @param id
	 *            Primary key for record.
	 * @param model
	 *            Values to pass through to view.
	 * @return View name to render
	 */
	@RequestMapping(value = "/edit/{id}", method = RequestMethod.GET)
	public String edit(@PathVariable int id, Model model) {
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product/edit request. Edit Action.");
			getLog().debug("Using id=[" + id + "]");
		}
		model.addAttribute(RequestConstants.REQUEST_HTTP_FORM_COMMAND,
				RequestConstants.REQUEST_HTTP_FORM_COMMAND_EDIT);
		model.addAttribute("productInfo", productService.get(id));
		model.addAttribute("message", "Editing  record");
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Edit Action.");
		}
		return "product/edit";
	}

	/**
	 * This action is responsible of saving a new record or update a record as
	 * well. <br/>
	 * Responds to the URI POST: {webapp}/product/update
	 * 
	 * @param info
	 *            Record information domain object to process.
	 * @param result
	 *            Result of binding HTTP objects to form instance fields.
	 * @param model
	 *            Values to pass through to view.
	 * @param request
	 *            The HTTP Request object.
	 * @return View name to render
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	public String update(@Valid ProductInfo info, BindingResult result,
			Model model, HttpServletRequest request) {
		String msg = "";
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product/update request. Update Action.");
		}
		if (result.hasErrors()) {
			return "product/edit";
		}
		// Decide if it is a new record or record update.
		if (info.getProductId() == 0) {
			if (getLog().isDebugEnabled()) {
				getLog().debug("Saving Record information (Create).");
			}
			try {
				productService.save(info);
				msg = composeMsgBoxMessage("Product created successfully.", "success", GlobalConstants.MSGBOX_DELAY_MILISECS);
				request.getSession(true).setAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG, msg);
			} catch (Exception e) {
				msg = composeMsgBoxMessage("Something went wrong creating the Product, ask your supplier for explanations.", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
				request.getSession(true).setAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG, msg);
			}
			
		} else {
			if (getLog().isDebugEnabled()) {
				getLog().debug("Saving Record information (Update).");
			}
			try {
				productService.update(info);
				msg = composeMsgBoxMessage("Product updated successfully.", "success", GlobalConstants.MSGBOX_DELAY_MILISECS);
				request.getSession(true).setAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG, msg);
			} catch (Exception e) {
				msg = composeMsgBoxMessage("Something went wrong updating the Product, ask your supplier for explanations.", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
				request.getSession(true).setAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG, msg);
			}
		}
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Update Action.");
		}
		return "redirect:";
	}

	/**
	 * Action to search records.<br/>
	 * Responds to the URI GET: {webapp}/product/search.
	 * 
	 * @param info
	 *            Record information domain object to process.
	 * @param result
	 *            Result of binding HTTP objects to form instance fields.
	 * @param model
	 *            Values to pass through to view.
	 * @param request
	 *            The HTTP Request object.
	 * @return View name to render
	 */
	@RequestMapping(value = "/search", method = RequestMethod.POST)
	public String search(@Valid ProductSearchInfo info, BindingResult result,
			Model model, HttpServletRequest request) {
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product/search request. Search Action.");
		}
		if (result.hasErrors()) {
			model.addAttribute("productListEmpty", false);
			return "product/index";
		}

		// Start logic to search records.
		List<ProductInfo> l = productService.findAllWith(info);
		model.addAttribute("productSearchInfo", info);
		model.addAttribute("productList", l);
		model.addAttribute("productListEmpty", l.isEmpty());
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Search Action.");
		}
		return "product/index";
	}

	/**
	 * Selected records to be deleted.<br/>
	 * Responds to the URI POST: {webapp}/product/delete
	 * 
	 * @param model
	 * @param request
	 *            The HTTP Request object.
	 * @return View name to render
	 */
	@RequestMapping(value = "/delete", method = RequestMethod.POST)
	public String delete(Model model, HttpServletRequest request) {
		if (getLog().isDebugEnabled()) {
			getLog().debug("Executing /product/delete request. Delete Action.");
		}
		// Delete logic here.
		String[] strValues = request.getParameterValues("recToDel");
		boolean isValidDelete = true;
		if (strValues != null && strValues.length > 0) {
			for (int i = 0; i < strValues.length; i++) {
				String s = strValues[i];
				try {
					productService.remove(Integer.parseInt(s));
				} catch (Exception e) {
					isValidDelete = false;
					getLog().info(s);
					e.printStackTrace();
				}
			}
		}
		String msg = "";
		if (isValidDelete) {
			msg = composeMsgBoxMessage(
					"Selected records successfully removed.", "success",
					GlobalConstants.MSGBOX_DELAY_MILISECS);
			/*
			 * getMessageSourceAccessor() .getMessage(
			 * "admon.usermnt.msg.valid.delete" );
			 */
		} else {
			msg = composeMsgBoxMessage(
					"There were some problems removing your selected records.",
					"error", GlobalConstants.MSGBOX_DELAY_MILISECS);
		}
		request.getSession(true).setAttribute(RequestConstants.REQUEST_HTTP_SESSION_MSG, msg);
		if (getLog().isDebugEnabled()) {
			getLog().debug("End Delete Action.");
		}
		return "redirect:";
	}

	/**
	 * Service injected by IoC.
	 * 
	 * @param productService
	 *            Service implementation for accessing products.
	 */
	@Autowired
	public void setProductService(IProductDataService productService) {
		this.productService = productService;
	}

	/**
	 * If the object reference dependency properties are not properly set, an
	 * exception is thrown indicating that they are required to be configured
	 * for the bean to work properly.
	 * 
	 * @throws Exception
	 */
	@Override
	public void afterPropertiesSet() throws Exception {
		 if (productService == null) {
             throw new Exception("productService property is not assigned.");
         }
	}
}

File: src/dotnet/DemoCSSln/DemoWeb/src/Controllers/ProductController.cs
/*==========================================================================*/
/* Source File:   PRODUCTCONTROLLER.CS                                      */
/* Description:   Controller Class for Products page.                       */
/* Author:        Carlos Adolfo Ortiz Quirós (COQ)                          */
/* Date:          Dec.03/2010                                               */
/* Last Modified: Dec,08/2010                                               */
/* Version:       1.3                                                       */
/* Copyright (c), 2010 CSoftZ                                               */
/*==========================================================================*/

/*===========================================================================
History
Dec.03/2010 COQ  File created.
============================================================================*/

using System;
using System.Web.Mvc;
using AutoMapper;
using CSoftZ.Demo.Common.Helper.Consts;
using CSoftZ.Demo.Domain;
using CSoftZ.Demo.Domain.View.Model;
using CSoftZ.Demo.Interface;
using Spring.Objects.Factory;

namespace CSoftZ.Demo.Web.Controllers {
    /// <summary>
    /// Controller class for Products Page.
    /// </summary>
    public class ProductController : AbstractSimpleController, IInitializingObject {
        // Injected by IoC. Access to Product services.
        private IProductDataService _productService;

        // Injected by IoC. Access to YesNo Services.
        private IYesNoValueService _yesNoValueService;

        // Injected by IoC. Access to ProductFilterCriteria Services.
        private IProductFilterCriteriaService _productFilterCriteriaService;

        // Injected by IoC. Access to PageSize Services.
        private IPageSizeValueService _pageSizeValueService;

        /// <summary>
        /// Define mapping for domain objects.
        /// </summary>
        private void SetupMappers() {
            Mapper.Reset();
            Mapper.CreateMap<ProductViewModel, ProductInfo>();
            Mapper.CreateMap<ProductInfo, ProductViewModel>();
        }

        ///  <summary>
        /// If the object reference dependency properties are not properly set, 
        /// an exception is thrown indicating that they are required to be 
        /// configured for the bean to work properly.
        /// </summary>
        /// <remarks></remarks>
        public void AfterPropertiesSet() {
            base.AfterPropertiesSet();
            if (_productService == null) {
                throw new Exception("productService property is not assigned.");
            }
            if (_yesNoValueService == null) {
                throw new Exception("yesNoValueService property is not assigned.");
            }
            if (_productFilterCriteriaService == null) {
                throw new Exception("productFilterCriteriaService property is not assigned.");
            }
            if (_pageSizeValueService == null) {
                throw new Exception("pageSizeValueService property is not assigned.");
            }
        }

        /// <summary>
        /// Welcome page for Products.
        /// </summary>
        /// <returns>Welcome page view.</returns>
        public ActionResult Index() {
            Boolean useSearch = false;
            ProductViewModel p = new ProductViewModel();
            ViewData[ViewConstants.VIEWDATA_FILTERCRITERIALIST] = _productFilterCriteriaService.GetAll();
            ViewData[ViewConstants.VIEWDATA_PAGESIZELIST] = _pageSizeValueService.GetAll();
            if (Request != null) {
                object sc = Request[GlobalConstants.REQUEST_SEARCH_CRITERIA_CB_HIDDEN];
                if (sc != null) {
                    string s = sc.ToString();
                    p.SearchCriteriaHidden = s;
                    p.SearchCriteria = Convert.ToInt32(s);
                }

                object sd = Request[GlobalConstants.REQUEST_SEARCH_CRITERIA_DATA_HIDDEN];
                if (sd != null) {
                    string s = sd.ToString();
                    p.SearchDataHidden = s;
                    p.SearchData = s;
                }

                object ps = Request[GlobalConstants.REQUEST_PAGE_SIZE];
                if (ps != null) {
                    string s = ps.ToString();
                    p.PageSize = Convert.ToInt32(s);
                }
                useSearch = true;

                // Now, if user has decided to use ALL, then resort to get all records instead.
                if (p.SearchCriteria == 0) {
                    useSearch = false;
                    p.SearchCriteriaHidden = "0";
                    p.SearchDataHidden = "";
                }
            }

            // Now it is time to infer if it is a full list or a searched list.
            if (useSearch) {
                p.ProductList = _productService.GetAll();
            }
            else {
                p.ProductList = _productService.GetAll();
            }
            return View(p);
        }

        /// <summary>
        ///  This is the entry point action to the Controller. When called it 
        ///  is because user activated the search criteria.</summary>
        /// <param name="m">Information to use.</param>
        /// <returns>Welcome page view.</returns>
        [HttpPost]
        public ActionResult Index(ProductViewModel m) {
            // Let//s first check if a DELETE is in progress.
            if (Request != null) {
                // NOTE: if HTML Input is BUTTON in ASP.NET MVC it is not sent
                // in a POST request, that is why I opted to used a hidden 
                // field to actually know if a DELETE button was used and it is
                // the one to test.
                var delRecs = Request[GlobalConstants.REQUEST_FORM_DELETE_RECORDS];
                var delRecsUsed = Request[GlobalConstants.REQUEST_FORM_DEL_RECS_USED];
                if (delRecsUsed != null) {
                    // Up to this point, a record deletion is in progress.
                    var recsToDelList = Request[GlobalConstants.REQUEST_FORM_REC_TO_DEL];
                    if (recsToDelList != null) {
                        var actionSuccess = true;
                        string s = recsToDelList.ToString();
                        String[] aiIds = s.Split(',');
                        foreach (var it in aiIds) {
                            int id = Convert.ToInt32(it);
                            try {
                                _productService.Remove(id);
                            }
                            catch (Exception) {
                                actionSuccess = false;
                            }
                        }

                        // Now compose message to view on screen.
                        if (actionSuccess) {
                            ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("One or more Products were successfully removed..", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
                        }
                        else {
                            ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("One or more Products were not successfully removed..", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
                        }
                    }
                } 
            }
            ProductViewModel p = new ProductViewModel();
            ViewData[ViewConstants.VIEWDATA_FILTERCRITERIALIST] = _productFilterCriteriaService.GetAll();
            ViewData[ViewConstants.VIEWDATA_PAGESIZELIST] = _pageSizeValueService.GetAll();
            if (p.PreserveCriteria == GlobalConstants.NO_VALUE) {
                p.SearchCriteria = m.SearchCriteria;
                p.SearchData = m.SearchData;
                p.SearchCriteriaHidden = Convert.ToString(m.SearchCriteria);
                p.SearchDataHidden = m.SearchData;
            }
            p.PageSize = m.PageSize;

            if (p.SearchCriteria == 0) {
                p.ProductList = _productService.GetAll();
            } else {
                p.ProductList = _productService.GetAll(); // findBy
            }
            return View(p);
        }

        /// <summary>
        /// Action to create an empty record from model.
        /// </summary>
        /// <returns>View with an empty product page</returns>
        public ActionResult Create() {
            ViewData[ViewConstants.VIEWDATA_YESNOVALUELIST] = _yesNoValueService.GetAll();
            ProductViewModel p = new ProductViewModel();
            return View(p);
        }

        /// <summary>
        /// Action to submit contents to be stored in database.
        /// </summary>
        /// <param name="m">Information to save</param>
        /// <returns></returns>
        [HttpPost]
        public ActionResult Create(ProductViewModel m) {
            SetupMappers();
            //ViewData[ViewConstants.VIEWDATA_MSGBOX] = "";
            //ViewData[ViewConstants.VIEWDATA_YESNOVALUELIST] = _yesNoValueService.GetAll();
            if (!ModelState.IsValid) {
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Uno o más campos son requeridos para completar el registro.", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
                return View();
            }

            // OK, form data is valid, now let's save it to database.
            try {
                ProductInfo info = Mapper.Map<ProductViewModel, ProductInfo>(m);
                _productService.Save(info);
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Product Successfully created.", "success", GlobalConstants.MSGBOX_DELAY_MILISECS);
            }
            catch (Exception) {
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Product not Successfully created.", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
            }
            //ProductViewModel p = new ProductViewModel();
           // return View(p);
            return RedirectToAction("Create");
        }

        /// <summary>
        /// Action to Edit a product record.
        /// </summary>
        /// <param name="id">The primary key to use.</param>
        /// <returns>View with record information to show</returns>
        public ActionResult Edit(int id) {
            ViewData[ViewConstants.VIEWDATA_YESNOVALUELIST] = _yesNoValueService.GetAll();
            ProductInfo info = _productService.Get(id);
            SetupMappers();
            ProductViewModel p = Mapper.Map<ProductInfo, ProductViewModel>(info);
            return View(p);
        }

        /// <summary>
        /// Action to submit contents to be stored in database.
        /// </summary>
        /// <param name="id">The primary key to use.</param>
        /// <param name="m">Information to save</param>
        /// <returns></returns>
        [HttpPost]
        public ActionResult Edit(ProductViewModel m) {
            ProductViewModel p = null;
            ViewData[ViewConstants.VIEWDATA_YESNOVALUELIST] = _yesNoValueService.GetAll();
            ViewData[ViewConstants.VIEWDATA_MSGBOX] = "";
            if (!ModelState.IsValid) {
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Uno o más campos son requeridos para completar el registro.", "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
                return View();
            }

            // OK, form data is valid, now let's save it to database.
            try {
                ProductInfo info = Mapper.Map<ProductViewModel, ProductInfo>(m);
                _productService.Update(info);
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Product Successfully edited.", "success", GlobalConstants.MSGBOX_DELAY_MILISECS);

                // Reload record to view again its data updated.
                info = _productService.Get(m.ProductId);
                SetupMappers();
                p = Mapper.Map<ProductInfo, ProductViewModel>(info);
            }
            catch (Exception e) {
                ViewData[ViewConstants.VIEWDATA_MSGBOX] = ComposeMsgBox("Product not Successfully edited." + e.Message, "error", GlobalConstants.MSGBOX_DELAY_MILISECS);
            }
            return View(p);
        }

        /// <summary>
        /// Injected by IoC. Access to Product Services.
        /// </summary>
        public IProductDataService ProductService {
            set { _productService = value; }
        }

        /// <summary>
        /// Injected by IoC. Access to YesNo Services.
        /// </summary>
        public IYesNoValueService YesNoValueService {
            set { _yesNoValueService = value; }
        }

        /// <summary>
        /// Injected by IoC. Access to ProductFilterCriteria Services.
        /// </summary>
        public IProductFilterCriteriaService ProductFilterCriteriaService {
            set { _productFilterCriteriaService = value; }
        }

        /// <summary>
        /// Injdected by IoC. Access PageSize services.
        /// </summary>
        public IPageSizeValueService PageSizeValueService {
            set { _pageSizeValueService = value; }
        }
    }
}
